### PR TITLE
Add Rest client middleware with unit tests

### PR DESCRIPTION
**Pull request by danielharasymiw-wf**
r: @tylertreat-wf 
o: @macleodbroad-wf @beaulyddon-wf @joestrach-wf

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Raven
Number of Findings: 0

File: rest/client.go [hunk 1]
```diff
@@ -33,10 +34,45 @@ const (
 	httpPut    = "PUT"
 )
 
+type InvocationHandler func(c *http.Client, method, url string, body interface{}, header http.Header) (*Response, error)
```
Comment (tylertreat-wf) [inline rest/client.go]: docstring

File: rest/client.go [hunk 2]
```diff
@@ -33,10 +34,45 @@ const (
...
+
+type ClientMiddleware func(InvocationHandler) InvocationHandler
```
Comment (tylertreat-wf) [inline rest/client.go]: docstring

File: rest/client.go [hunk 3]
```diff
@@ -33,10 +34,45 @@ const (
...
+
+type HttpClient interface {
```
Comment (tylertreat-wf) [inline rest/client.go]: docstring

File: rest/client.go [hunk 4]
```diff
@@ -33,10 +34,45 @@ const (
...
+	Do(req *http.Request) (resp *http.Response, err error)
+	Get(url string) (resp *http.Response, err error)
+	Post(url string, bodyType string, body io.Reader) (resp *http.Response, err error)
+	PostForm(url string, data url.Values) (resp *http.Response, err error)
+	Head(url string) (resp *http.Response, err error)
+}
+
+type RestClient interface {
```
Comment (tylertreat-wf) [inline rest/client.go]: docstring

**Comment by tylertreat-wf** [comment]
Could use some docstrings but otherwise lgtm.

File: rest/client.go [hunk 5]
```diff
@@ -34,10 +34,13 @@ const (
 	httpPut    = "PUT"
 )
 
+// The type that is the function that is to be wrapped by the ClientMiddleware
 type InvocationHandler func(c *http.Client, method, url string, body interface{}, header http.Header) (*Response, error)
```
Comment (tylertreat-wf) [inline rest/client.go]: This is a nit, but Go docstrings should start with the name of the type/function/struct/etc. https://blog.golang.org/godoc-documenting-go-code

Comment (danielharasymiw-wf) [inline rest/client.go]: Thanks for the link, I've updated the docstrings and will try and keep it in mind for future use.

**Comment by joestrach-wf** [comment]
+1 looks great

**Comment by beaulyddon-wf** [comment]
+1

**Comment by tylertreat-wf** [comment]
+1

**Comment by teresarevious-wf** [comment]
RM +1

 verified dependencies manually

**Comment by rmconsole-wf** [comment]
QA review can be found here: https://jira.atl.workiva.net/browse/RM-16257

**Pull request merged**